drivers/gpu/drm/virtio/virtgpu_prime.c

```diff
@@ -155,16 +155,101 @@ struct drm_gem_object *virtgpu_gem_prime_import(struct drm_device *dev,
 			 * refcount on gem itself instead of f_count of dmabuf.
 			 */
 			drm_gem_object_get(obj);
+
 			return obj;
 		}
 	}
 
 	return drm_gem_prime_import(dev, buf);
 }
 
+static int virtio_gpu_sgt_to_mem_entry(struct virtio_gpu_device *vgdev,
+		                  struct sg_table *table,
+                                  struct virtio_gpu_mem_entry **ents,
+				  unsigned int *nents)
+{
+
+	struct scatterlist *sg;
+	int si;
+
+	bool use_dma_api = !virtio_has_dma_quirk(vgdev->vdev);
+	if (use_dma_api)
+		*nents = table->nents;
+	else
+		*nents = table->orig_nents;
+
+	*ents = kvmalloc_array(*nents,
+			       sizeof(struct virtio_gpu_mem_entry),
+			       GFP_KERNEL);
+	if (!(*ents)) {
+		DRM_ERROR("failed to allocate ent list\n");
+		return -ENOMEM;
+	}
+
+	if (use_dma_api) {
+		for_each_sgtable_dma_sg(table, sg, si) {
+			(*ents)[si].addr = cpu_to_le64(sg_dma_address(sg));
+			(*ents)[si].length = cpu_to_le32(sg_dma_len(sg));
+			(*ents)[si].padding = 0;
+		}
+	} else {
+		for_each_sgtable_sg(table, sg, si) {
+			(*ents)[si].addr = cpu_to_le64(sg_phys(sg));
+			(*ents)[si].length = cpu_to_le32(sg->length);
+			(*ents)[si].padding = 0;
+		}
+	}
+
+	return 0;
+
+}
+
 struct drm_gem_object *virtgpu_gem_prime_import_sg_table(
 	struct drm_device *dev, struct dma_buf_attachment *attach,
 	struct sg_table *table)
 {
-	return ERR_PTR(-ENODEV);
+	size_t size = PAGE_ALIGN(attach->dmabuf->size);
```

Comment:
do we need to validate attach and attach->dmabuf

Reply:
the function virtgpu_gem_prime_import_sg_table is the callback function virtio gpu driver register to the drm framework (.gem_prime_import_sg_table defined in struct drm_driver_driver in the file virtgpu_drv.c). and drm framework will call this function in the file drm_prime.c, and before the calling , the caller ensure the attach is validated.